### PR TITLE
Invoice item

### PR DESCRIPTION
**Pull request by lgsriclas**
Please take a look at the invoice item files   Thanks!

### CODE REVIEW

File: lib/invoice_item.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    @item_id = data[:item_id].to_i
+    @invoice_id = data[:invoice_id].to_i
+    @quantity = data[:quantity].to_i
+    @unit_price = data[:unit_price].to_f
+    @created_at = data[:created_at]
+    @updated_at = data[:updated_at]
+  end
+
+  def unit_price_to_dollars
+    @unit_price.to_f
```
Comment (michab17) [inline lib/invoice_item.rb]: This may need to have a $ in front of it

File: lib/invoice_item_repository.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def all
+    result = @rows.map do |row|
+      InvoiceItem.new(row)
+    end
+  end
+
+  def find_by_id(id)
+    result = @all.find do |row|
```
Comment (michab17) [inline lib/invoice_item_repository.rb:17]: I dont believe you need to set this equal to result and then return result, it should output the same without it

File: spec/invoice_item_repository_spec.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      expect(iir).to be_an_instance_of InvoiceItemRepository
+    end
+  end
+
+  describe 'all' do
+    it 'returns an array of all invoice item instances' do
+      path = './data/invoice_items.csv'
+      iir = InvoiceItemRepository.new(path)
+
+      expect(iir.all).to be_an Array
```
Comment (michab17) [inline spec/invoice_item_repository_spec.rb:19]: You could add another test here testing the length of the array to see if all of them are included

**Review by michab17 — COMMENTED** [review]
I added some changes to look at!

File: lib/invoice_item.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  def initialize(data)
+    @id = data[:id].to_i
+    @item_id = data[:item_id].to_i
+    @invoice_id = data[:invoice_id].to_i
+    @quantity = data[:quantity].to_i
+    @unit_price = data[:unit_price].to_f
+    @created_at = data[:created_at]
+    @updated_at = data[:updated_at]
+  end
```
Comment (samueldevine) [inline lib/invoice_item.rb:20]: these could be lined up to look cleaner but nbd

File: lib/invoice_item_repository.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+class InvoiceItemRepository
+  def initialize(path)
+    @path = path
+    @rows = CSV.read(@path, headers: true, header_converters: :symbol)
+    @all = all
+  end
+
+  def all
+    result = @rows.map do |row|
```
Comment (samueldevine) [inline lib/invoice_item_repository.rb]: ^

File: lib/invoice_item_repository.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      row.id == id
+    end
+    result
+  end
+
+  def find_all_by_item_id(item_id)
+    result = @all.find_all do |row|
+      row.item_id == item_id
+    end
+    result
```
Comment (samueldevine) [inline lib/invoice_item_repository.rb]: similar to above, I think the `result` on lines 24 and 27 is unnecessary; `#find_all` should return the collection by itself

File: lib/invoice_item_repository.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def find_all_by_invoice_id(invoice_id)
+    result = @all.find_all do |row|
+      row.invoice_id == invoice_id
+    end
+    result
```
Comment (samueldevine) [inline lib/invoice_item_repository.rb]: same here with the results

File: lib/invoice_item_repository.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def create(attributes)
+    attributes[:id] = @all.last.id.to_i + 1
```
Comment (samueldevine) [inline lib/invoice_item_repository.rb]: don't need the `to_i` if you are already converting the `id` during instantiation

File: lib/invoice_item_repository.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def update(id, attributes)
+    invoice_item_to_update = find_by_id(id)
+    if attributes[:quantity] != nil
+      invoice_item_to_update.quantity = attributes[:quantity]
+    end
+    if attributes[:unit_price] != nil
+      invoice_item_to_update.unit_price = attributes[:unit_price]
+    end
```
Comment (samueldevine) [inline lib/invoice_item_repository.rb:47]: nice catch checking for nil values here!

I handled this using an **inline conditional** to eliminate a few lines of code, but it also made my lines really long. 
Inline conditionals would look like this:

`invoice_item_to_update.quantity = attributes[:quantity] if attributes[:quantity]`

The `if` essentially checks if the value exists or if it's `nil`, and then assigns the value to the variable if it it returns `truthy` (i.e. if it exists). If your lines get too long you can rename `invoice_item_to_update` and/or assign `attributes[:quantity]` to a variable, etc.

File: spec/invoice_item_repository_spec.rb [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    it 'creates a new instance of InvoiceItem' do
+      path = './data/invoice_items.csv'
+      iir = InvoiceItemRepository.new(path)
+
+      attributes = {
+        :id          => 6,
+        :item_id     => 7,
+        :invoice_id  => 8,
+        :quantity    => 1,
+        :unit_price  => 1099,
```
Comment (samueldevine) [inline spec/invoice_item_repository_spec.rb]: to clarify: the Ruby 1.9 hash syntax is 
{id: 6, item_id: 8, ...}
vs.
{:id => 6, :item_id => 8, ...}
minor style things from hound

File: spec/invoice_item_spec.rb [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+      expect(@ii.created_at).to eq(Time.now.round(1))
+    end
+
+    it 'has a updated at time' do
+      expect(@ii.updated_at).to eq(Time.now.round(1))
+    end
+
+    it 'returns the price of invoice items' do
+      expect(@ii.unit_price_to_dollars).to be_a(Float)
+    end
```
Comment (samueldevine) [inline spec/invoice_item_spec.rb:52]: we could move all of the attribute tests into one `it` block to save some space; they're all essentially testing the same thing

**Review by samueldevine — CHANGES_REQUESTED** [review]
looks great overall! left a few comments to update before we merge :)

**Comment by haewonito** [comment]
I don't see anything to add to what Micha and Sam have already pointed out.

**Pull request merged**